### PR TITLE
Retroactive Implements - Scala side

### PR DESCRIPTION
**Pull request by akrmn**

### CODE REVIEW

File: daml-lf/archive/src/test/scala/com/digitalasset/daml/lf/archive/DecodeV1Spec.scala [hunk 1]
```diff
@@ -1050,6 +1050,7 @@ class DecodeV1Spec
                 Map(),
                 Map(),
                 EPrimCon(PCUnit),
+                Map(),
```
Comment (cocreature) [inline daml-lf/archive/src/test/scala/com/digitalasset/daml/lf/archive/DecodeV1Spec.scala:1053]: Are you handling tests separately? If so, maybe worth tracking that somewhere? If not, I think you’re missing some tests for this.

**Review by cocreature — APPROVED** [review]
Nice, thank you!

File: daml-lf/parser/src/main/scala/com/digitalasset/daml/lf/testing/parser/AstRewriter.scala [hunk 2]
```diff
@@ -342,15 +342,39 @@ private[daml] class AstRewriter(
...
     x match {
-      case DefInterface(requires, param, choices, methods, precond) =>
+      case DefInterface(requires, param, choices, methods, precond, coImplements) =>
         DefInterface(
           requires.map(apply(_)),
           param,
           choices.transform((_, v) => apply(v)),
           methods.transform((_, v) => apply(v)),
           apply(precond),
+          coImplements.map({ case (t, x) => (apply(t), apply(x)) }),
```
Comment (remyhaemmerle-da) [inline daml-lf/parser/src/main/scala/com/digitalasset/daml/lf/testing/parser/AstRewriter.scala:377]: purely cosmetic

```suggestion
          coImplements.map{ case (t, x) => (apply(t), apply(x)) },
```

File: daml-lf/archive/src/test/scala/com/digitalasset/daml/lf/archive/DecodeV1Spec.scala (hunk 1 above)
Comment (remyhaemmerle-da) [reply daml-lf/archive/src/test/scala/com/digitalasset/daml/lf/archive/DecodeV1Spec.scala:1053]: We are indeed missing tests ;)

**Review by remyhaemmerle-da — APPROVED** [review]
Nice work. Thanks a lot

File: daml-lf/archive/src/test/scala/com/digitalasset/daml/lf/archive/DecodeV1Spec.scala (hunk 1 above)
Comment (akrmn) [reply daml-lf/archive/src/test/scala/com/digitalasset/daml/lf/archive/DecodeV1Spec.scala:1053]: ah yes, it seems our only scala decoder test for interfaces just tests the version; I'll add a couple more tests for the components of an interface definition

File: daml-lf/parser/src/main/scala/com/digitalasset/daml/lf/testing/parser/AstRewriter.scala (hunk 2 above)
Comment (akrmn) [reply daml-lf/parser/src/main/scala/com/digitalasset/daml/lf/testing/parser/AstRewriter.scala:377]: ah nice, I'd copied this part from the version of `apply` for `Template`, I'll fix that one too.

File: daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/iterable/TypeIterable.scala [hunk 3]
```diff
@@ -255,13 +255,27 @@ private[validation] object TypeIterable {
         iterator(value)
     }
 
+  private[validation] def iterator(coImpl: InterfaceCoImplements): Iterator[Type] =
+    coImpl match {
+      case InterfaceCoImplements(template, methods) =>
+        Iterator(TTyCon(template)) ++
+          methods.values.flatMap(iterator)
```
Comment (remyhaemmerle-da) [inline daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/iterable/TypeIterable.scala:262]:
```suggestion 
          methods.values.iterator.flatMap(iterator)
```
Without the iterator, the type of  `methods.values.flatMap(iterator)` produces a `List` which is pretty terrible in most of the use cases `TypeIterable` which are recursive. Here we really want to to the concatenation lazily, so we need to insert a `.view` (which is pure) or a `.iterator` between the `values` and the `flatMap`.  In this case, I would favor `.iterator` as the return type is also an `iterator`

File: daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/iterable/TypeIterable.scala [hunk 4]
```diff
@@ -255,13 +255,27 @@ private[validation] object TypeIterable {
...
   private[validation] def iterator(interface: DefInterface): Iterator[Type] =
     interface match {
-      case DefInterface(requires, _, choices, methods, precond) =>
+      case DefInterface(requires, _, choices, methods, precond, coImplements) =>
         requires.iterator.map(TTyCon) ++
           iterator(precond) ++
           choices.values.iterator.flatMap(iterator) ++
-          methods.values.iterator.flatMap(iterator)
+          methods.values.iterator.flatMap(iterator) ++
+          coImplements.values.flatMap(iterator)
```
Comment (remyhaemmerle-da) [inline daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/iterable/TypeIterable.scala:278]:
```suggestion 
          methods.values.iterator.flatMap(iterator) ++
          coImplements.values.iterator.flatMap(iterator)
```

File: daml-lf/archive/src/test/scala/com/digitalasset/daml/lf/archive/DecodeV1Spec.scala [hunk 5]
```diff
@@ -1072,6 +1084,137 @@ class DecodeV1Spec
...
+        .setTyconInternedDname(1)
+        .setParamInternedStr(0)
+        .setPrecond(unitExpr)
+        .build()
+
+      val emptyDefInterfaceScala =
+        GenDefInterface(
+          Set.empty,
+          Ref.IdString.Name.assertFromString("this"),
+          Map(),
```
Comment (remyhaemmerle-da) [inline daml-lf/archive/src/test/scala/com/digitalasset/daml/lf/archive/DecodeV1Spec.scala:1105]: In general I favor Map.empty.

**Review by remyhaemmerle-da — COMMENTED** [review]
Very Nice. 

There is a minor issue in `TypeIterable` though. 

Please link the issue and update the progress in the issue.

**Pull request merged**